### PR TITLE
Add Fuchsia support and CI

### PR DESCRIPTION
**Pull request by tamird**
In general, feature support is identical to Linux.

Updates #78.

### CODE REVIEW

File: src/sys/unix.rs [hunk 1]
```diff
@@ -519,15 +526,14 @@ fn recvmsg(
...
-    let mut msg = libc::msghdr {
-        msg_name: msg_name.cast(),
-        msg_namelen,
-        msg_iov: bufs.as_mut_ptr().cast(),
-        msg_iovlen: min(bufs.len(), IovLen::MAX as usize) as IovLen,
-        msg_control: ptr::null_mut(),
-        msg_controllen: 0,
-        msg_flags: 0,
-    };
+    let mut msg: libc::msghdr = unsafe { mem::zeroed() };
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:529]: Can you revert back to creating the structure directly, it makes a bit more awkward with differences between OSs I know, but it forces you to check the structure when porting to a new OS (like what happend here).

File: src/sys/unix.rs [hunk 2]
```diff
@@ -578,19 +584,18 @@ fn sendmsg(
...
-        msg_name: (msg_name as *mut sockaddr_storage).cast(),
-        msg_namelen,
-        // Safety: Same as above about `*const` -> `*mut`.
-        msg_iov: bufs.as_ptr() as *mut _,
-        msg_iovlen: min(bufs.len(), IovLen::MAX as usize) as IovLen,
-        msg_control: ptr::null_mut(),
-        msg_controllen: 0,
-        msg_flags: 0,
-    };
+    let mut msg: libc::msghdr = unsafe { mem::zeroed() };
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:587]: Same as above.

**Review by Thomasdezeeuw — APPROVED** [review]
Two small things, but LGTM otherwise.

**Comment by Thomasdezeeuw** [comment]
@tamird is there manual online for libc/POSIX complicated system call interface for Fuchsia? A link I can add to https://github.com/rust-lang/socket2/pull/154?

**Comment by Thomasdezeeuw** [comment]
For myself, update https://github.com/rust-lang/socket2/issues/78.

File: src/sys/unix.rs (hunk 1 above)
Comment (tamird) [reply src/sys/unix.rs:529]: The trouble is that this structure has private padding fields on Fuchsia, and can't be instantiated directly.

Shall I add a comment? Is there a better solution?

**Comment by tamird** [comment]
> @tamird is there manual online for libc/POSIX complicated system call interface for Fuchsia? A link I can add to #154?

Not that I'm aware of at this time. Fuchsia's libc is derived from musl, so it is largely in line with Linux (or is intended to be, in areas where it is not yet so).

File: src/sys/unix.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply src/sys/unix.rs:529]: Does something like the following work?
```
msghdr {
    // fields ...
    .. Default::default(),
}
```

Or does `msghdr` need to implement `Default` for that? I forgot...

If it's not possible, keep the change as is but with a comment indeed. And can you remove the `msg_controllen` and `msg_flags` fields as there already zero.

Comment (tamird) [reply src/sys/unix.rs:530]: Yeah, it has to impl `Default`. Added a comment and removed zero initialization.

File: src/sys/unix.rs (hunk 2 above)
Comment (tamird) [reply src/sys/unix.rs:586]: Done.

**Comment by Thomasdezeeuw** [comment]
Thanks @tamird.

**Comment by Thomasdezeeuw** [comment]
Hey @tamird, apparently Fuchsia doesn't yet work. I tried checking it locally and found the following errors:

```log
error[E0425]: cannot find value `IFNAMSIZ` in crate `libc`
   --> src/sys/unix.rs:934:46
    |
934 |         let mut buf: [MaybeUninit<u8>; libc::IFNAMSIZ] =
    |                                              ^^^^^^^^ not found in `libc`

error[E0425]: cannot find value `IFNAMSIZ` in crate `libc`
   --> src/sys/unix.rs:935:60
    |
935 |             unsafe { MaybeUninit::<[MaybeUninit<u8>; libc::IFNAMSIZ]>::uninit().assume_init() };
    |                                                            ^^^^^^^^ not found in `libc`

error[E0425]: cannot find value `IP_FREEBIND` in crate `libc`
    --> src/sys/unix.rs:1049:69
     |
1049 |             getsockopt::<c_int>(self.inner, libc::SOL_SOCKET, libc::IP_FREEBIND)
     |                                                                     ^^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find value `IP_FREEBIND` in crate `libc`
    --> src/sys/unix.rs:1072:23
     |
1072 |                 libc::IP_FREEBIND,
     |                       ^^^^^^^^^^^ not found in `libc`

```

So `IFNAMSIZ` and `IP_FREEBIND` aren't defined. It's also a problem in the CI, but it passed anyway (now issue #169). Are these two constant available already in Fuchsia?

**Comment by tamird** [comment]
That's odd! Yeah, those constants are defined in our libc:

`IFNAMSIZ`: https://cs.opensource.google/fuchsia/fuchsia/+/master:zircon/third_party/ulib/musl/include/net/if.h;l=72;drc=bdf60afa9193c8a9190ee6bb6b5a0a11ad8876b6

`IP_FREEBIND`: https://cs.opensource.google/fuchsia/fuchsia/+/master:zircon/third_party/ulib/musl/include/netinet/in.h;l=196;drc=bdf60afa9193c8a9190ee6bb6b5a0a11ad8876b6

but apparently not in rust-lang/libc:

`IFNAMSIZ` should be here: https://github.com/rust-lang/libc/blob/c9fb51bc650427fe22dcf34b503f6aa6fc909c07/src/fuchsia/mod.rs#L1329 -- sent https://github.com/rust-lang/libc/pull/2010.

`IP_FREEBIND` should be here: https://github.com/rust-lang/libc/blob/c9fb51bc650427fe22dcf34b503f6aa6fc909c07/src/fuchsia/mod.rs#L1765 -- sent https://github.com/rust-lang/libc/pull/2011.

**Pull request merged**